### DISCUSSION TITLE
Convert some jupyter notebook functionality to standalone scripts

### DISCUSSION THREAD

**Discussion by zhonge** [zhonge/cryodrgn#89] [Ideas]
This is an often requested feature especially from those with computing setups that make it hard to use jupyter notebooks.  Some of the interactive widgets in the jupyter notebook also tend to fail above 500k particles. 

Some functionality that would be easy to transfer:
* plot by CTF, plot by pose, plot by particle id
* filtering by z-score
* combining particle selections

**Comment by Guillawme** [comment]
Filtering by ranges of UMAP values would also be super helpful! I am dealing right now with ~700k particles and the interactive plot indeed chokes and never works (I have to kill the frozen notebook every time I run this cell inadvertently).

I think the easiest user interface would be to provide ranges of UMAP values that bracket the cluster of particles the user wants to **keep**, maybe something like `--umap1-range 2 5 --umap2-range 3 6`, because these ranges are easy to eyeball on the UMAP plot. It would also be great if this command could generate a new UMAP plot (also the hexbin version) with the selected particles in a different color, so the user can confirm that the range they eyeballed in step one is what they really want (and can easily rerun the command with a refined range if the plot shows that the selection is too broad or too narrow).

**Reply by olibclarke** [reply]
Agree this would be very helpful! Alternatively/additionally, perhaps it would be possible to have the widget to show a random subset, if the particle number is above a certain threshold? A random subset of say 50k particles, as used by `analyze_convergence.py`, should be fine for visually selecting a group of particles I should think, and then maybe it would behave better?

**Reply by adriangl97** [reply]
Hi @Guillawme ,
There is a way to do this in the Jupyter notebook. I have sent a pull request #313 for implementation, but in the meantime, just paste this code in the second cell of the interactive selection. Then modify as needed. Run the next cells to get the coloring of the selection.

```
# 1 selection
ind_selected1 = df.index[(df['UMAP1'] >= -5) & (df['UMAP1'] <= 5) & (df['UMAP2'] >= -5) & (df['UMAP2'] <= 5)]
ind_selected1 = np.array(ind_selected1)
ind_selected = ind_selected1
# 2 selections
#ind_selected2 = df.index[(df['UMAP1'] >= -5) & (df['UMAP1'] <= 5) & (df['UMAP2'] >= -5) & (df['UMAP2'] <= 5)]
#ind_selected2 = np.array(ind_selected2)
#ind_selected = np.append(ind_selected1, ind_selected2)
#ind_selected = np.unique(ind_selected)
# 3 selections
#ind_selected3 = df.index[(df['UMAP1'] >= -5) & (df['UMAP1'] <= 5) & (df['UMAP2'] >= -5) & (df['UMAP2'] <= 5)]
#ind_selected3 = np.array(ind_selected3)
#ind_selected = np.append(ind_selected, ind_selected3)
#ind_selected = np.unique(ind_selected)

ind_selected_not = invert_selection(ind_selected)

print('Selected indices:')
print(ind_selected)
print('Number of selected points:')
print(len(ind_selected))
print('Number of unselected points:')
print(len(ind_selected_not))
```

**Reply by Guillawme** [reply]
Awesome, thank you for taking the time to figure this out, and thank you for sharing!
